design/authentication-flow-usability.md

```diff
@@ -0,0 +1,322 @@
+I think it may be similar for other server adapters (node.js, gatekeeper etc) but will be good to doublecheck.
+For the keycloak.js, I think "login_required" will work fine and automatically redirect to KC, but for the applications with "check_sso" or without any "onLoad" value used,
+(For example _js-console_ application from the deprecated keycloak-examples), the adapter won't automatically redirect to Keycloak. So some changes in the adapter
+may be needed to achieve this.
+
+4. For SAML, will be good to doublecheck the behaviour. I suppose that SAML adapters are server-side and hence is not an
+issue to redirect back to the application.
+
+
+
+### Advanced approach
+
+Advanced approach would be more work, but will help with some more advanced use-cases described in the [minimalistic approach](##minimal-approach)
+and would mean some more improvements regarding usability of authentication flows and memory consumption. Especially:
+
+* Mitigate the usability issues above and improve usability of authentication in general
+
+* Minimalize "error" redirects to the application
+
+* Minimalize impact on the memory and ensure that most of the state is saved in the browser
+
+The disadvantage is more refactoring and more work needed and the risk of some more regressions.
+
+
+#### Advanced approach - solution proposal
+
+From the high level point of view, advanced approach would mean that "authentication data" (See [above](#authentication-sessions---current-model) for what it means) are
+ideally shared per whole browser - per all browser tabs. Client data are not tracked anymore in the authentication session on
+the server, but in the browser - either in the request parameters sent across all requests or per browser cookie.
+
+1. In addition to the AUTH_SESSION_ID cookie, we will introduce another cookie like AUTH_TIMESTAMP. Cookie will contain
+timestamp in the integer format. This cookie will be updated after every action triggered in the authenticators.
+
+2. Every page with the authenticators will contain some small JS code, which will periodically check the AUTH_TIMESTAMP cookie
+mentioned above. In case that it recognizes the change, the page will be automatically refreshed and moved to the latest state.
+For example if user confirms username/password in browser tab1, which moved to the TOTP screen, then tab2 will be moved in the
+background to the TOTP screen as well.
+In details, when javascript recognize the change in the timestamp cookie, then
+tab2 will:
+    * Send the browser "refresh" request
+    
+    * Authentication session will be shared per whole browser (with some exceptions described below in section [different flows in same browser](#different-flows-in-same-browser).
+    So on KC side, we will check the latest state from AuthenticationSession and update the page accordingly including URLs. This will allow all browser tabs will be updated to TOTP after confirm username/password in tab1
```

Comment:
I agree that first priority is to get rid of "You are already logged-in" . Also agree with the point of more uneccessary browser requests. However I think that manual refreshes is not very good. That would actually mean that user will in most cases still see the username/password screen and fill the username/password as most of users don't do browser refresh.

So maybe the JS can refresh the page in case that particular browser tab is clicked and gains focus? We can always refresh the page when gaining focus, which means that we won't need the timestamp cookie at all then.

The Idp flow or client-specific flow shouldn't be an issue. As mentioned below, there will be still more authentication sessions in the rootAuthSession to handle those cases. Example:
- There are 3 browser tabs. Tabs 1 and 2 are using flow1 (default realm flow). Tab 3 is using some client specific flow - flow2.
- User fills username/password on tab 1. He is moved to step 2 (TOTP screen)
- User clicks on tab 2. JS will refresh the page and user will be moved to TOTP screen
- User clicks on tab 3. JS will refresh the page, but in the flow 2 (client specific flow), user didn't anything yet. So refresh will just keep user at step 1 of the flow 2.

Reply:
> So maybe the JS can refresh the page in case that particular browser tab is clicked and gains focus? We can always refresh the page when gaining focus, which means that we won't need the timestamp cookie at all then.

+1 to this approach. When the user clicks the tab, we only need to put some notification to the user that the page is being refreshed so that they don't fill and submit a form in the meantime (could be just a shading full screen div or whatever)

Reply:
I've changed the design per this.